### PR TITLE
Make UserAgent request header optional

### PR DESCRIPTION
**Pull request by KirovAir**
As per title: Make UserAgent request header optional. Or even opt-in. For a lot of projects it's redundant and it can interfere with multi threaded applications where the HttpClient is shared. The Add() will cause a race condition error.

### CODE REVIEW

File: src/GraphQL.Client/GraphQLHttpClientOptions.cs [hunk 1]
```diff
@@ -67,5 +67,11 @@ public class GraphQLHttpClientOptions
         /// See https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md#gql_connection_init.
         /// </summary>
         public Func<GraphQLHttpClientOptions, object?> ConfigureWebSocketConnectionInitPayload { get; set; } = options => null;
+
+
+        /// <summary>
+        /// Add the assembly name as UserAgent request header if none is set.
+        /// </summary>
+        public bool AddDefaultUserAgentRequestHeader { get; set; } = true;
```
Comment (sungam3r) [inline src/GraphQL.Client/GraphQLHttpClientOptions.cs:75]: I would suggest to not touch `DefaultRequestHeaders` at all:

1. Change new option to something like 
```
public Func<GraphQLClientHttpOptions, ProductInfoHeaderValue> DefaultUserAgentRequestHeader { get; set; } = _ => new ProductInfoHeaderValue(GetType().Assembly.GetName().Name, GetType().Assembly.GetName().Version.ToString())
```

2. Remove
```
 if (!HttpClient.DefaultRequestHeaders.UserAgent.Any())
     HttpClient.DefaultRequestHeaders.UserAgent.Add(ProductInfoHeaderValue(GetType().Assembly.GetName().Name, GetType().Assembly.GetName().Version.ToString()));
```

3. Add 
```
            var userAgent = options.DefaultUserAgentRequestHeader(options);
            if (userAgent != null)
                message.Headers.UserAgent.Add(userAgent);
```

into `GraphQLHttpRequest.ToHttpRequestMessage`.

**Comment by KirovAir** [comment]
I've added the requested changes. :)
Do you think the function wrapper is needed? I have not so much experience with them so I'm not entirely sure if the current implementation is correct.

**Comment by sungam3r** [comment]
function wrapper?

**Comment by KirovAir** [comment]
It was a bit late, I ment the delegate around the option. Anyway this current implementation works as intended.

**Comment by sungam3r** [comment]
Conflicts 😕

**Comment by rose-a** [comment]
@KirovAir pls rebase on current master

**Comment by KirovAir** [comment]
Done!

File: src/GraphQL.Client/GraphQLHttpClientOptions.cs [hunk 2]
```diff
@@ -67,5 +67,11 @@ public class GraphQLHttpClientOptions
         /// See https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md#gql_connection_init.
         /// </summary>
         public Func<GraphQLHttpClientOptions, object?> ConfigureWebSocketConnectionInitPayload { get; set; } = options => null;
+
+        /// <summary>
+        /// The default user agent request header.
+        /// Default to the GraphQL client assembly.
+        /// </summary>
+        public Func<GraphQLHttpClientOptions, ProductInfoHeaderValue> DefaultUserAgentRequestHeader { get; set; } = _ => new ProductInfoHeaderValue(typeof(GraphQLHttpClient).Assembly.GetName().Name, typeof(GraphQLHttpClient).Assembly.GetName().Version.ToString());
```
Comment (rose-a) [inline src/GraphQL.Client/GraphQLHttpClientOptions.cs:75]: > I've added the requested changes. :) Do you think the function wrapper is needed? I have not so much experience with them so I'm not entirely sure if the current implementation is correct.

I don't think a function delegate is appropriate here, as the user agent request header sounds like something static across a single application...

I'm in favour of a regular property...

```suggestion
        public ProductInfoHeaderValue DefaultUserAgentRequestHeader { get; set; } = new ProductInfoHeaderValue(typeof(GraphQLHttpClient).Assembly.GetName().Name, typeof(GraphQLHttpClient).Assembly.GetName().Version.ToString());
```

File: src/GraphQL.Client/GraphQLHttpRequest.cs [hunk 3]
```diff
@@ -41,6 +41,10 @@ public virtual HttpRequestMessage ToHttpRequestMessage(GraphQLHttpClientOptions
                 Content = new StringContent(serializer.SerializeToString(this), Encoding.UTF8, options.MediaType)
             };
 
+            var userAgentOption = options.DefaultUserAgentRequestHeader;
```
Comment (rose-a) [inline src/GraphQL.Client/GraphQLHttpRequest.cs:44]:
```suggestion
```

File: src/GraphQL.Client/GraphQLHttpRequest.cs [hunk 4]
```diff
@@ -41,6 +41,10 @@ public virtual HttpRequestMessage ToHttpRequestMessage(GraphQLHttpClientOptions
...
+            if (userAgentOption != null)
```
Comment (rose-a) [inline src/GraphQL.Client/GraphQLHttpRequest.cs:45]:
```suggestion
            if (options.DefaultUserAgentRequestHeader != null)
```

File: src/GraphQL.Client/GraphQLHttpRequest.cs [hunk 5]
```diff
@@ -41,6 +41,10 @@ public virtual HttpRequestMessage ToHttpRequestMessage(GraphQLHttpClientOptions
...
+                message.Headers.UserAgent.Add(userAgentOption(options));
```
Comment (rose-a) [inline src/GraphQL.Client/GraphQLHttpRequest.cs:46]:
```suggestion
                message.Headers.UserAgent.Add(options.DefaultUserAgentRequestHeader);
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 6]
```diff
@@ -180,6 +180,9 @@ public async void PreprocessHttpRequestMessageIsCalled()
             };
 
             var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
+            var userAgentOption = StarWarsClient.Options.DefaultUserAgentRequestHeader;
```
Comment (rose-a) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:183]: Why are you doing this here? This does not test the new functionality!

Please create a new test case to prove that the user agent header is correctly modified.

**Review by rose-a — CHANGES_REQUESTED** [review]

File: src/GraphQL.Client/GraphQLHttpRequest.cs [hunk 7]
```diff
@@ -43,6 +43,9 @@ public virtual HttpRequestMessage ToHttpRequestMessage(GraphQLHttpClientOptions
         message.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
         message.Headers.AcceptCharset.Add(new StringWithQualityHeaderValue("utf-8"));
 
+            if (options.DefaultUserAgentRequestHeader != null)
+                message.Headers.UserAgent.Add(options.DefaultUserAgentRequestHeader);
```
Comment (sungam3r) [inline src/GraphQL.Client/GraphQLHttpRequest.cs:47]:
```suggestion
        if (options.DefaultUserAgentRequestHeader != null)
            message.Headers.UserAgent.Add(options.DefaultUserAgentRequestHeader);
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 8]
```diff
@@ -152,6 +152,16 @@ public async void SendMutationFact()
             }
         });
 
+
```
Comment (sungam3r) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:155]:
```suggestion
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 9]
```diff
@@ -152,6 +152,15 @@ public async void SendMutationFact()
             }
         });
 
+        var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
+        var userAgentOption = StarWarsClient.Options.DefaultUserAgentRequestHeader;
+        if (userAgentOption != null)
+            defaultHeaders.UserAgent.Add(userAgentOption(StarWarsClient.Options));
```
Comment (sungam3r) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:158]:
```suggestion
        var userAgent = StarWarsClient.Options.DefaultUserAgentRequestHeader;
        if (userAgent != null)
            defaultHeaders.UserAgent.Add(userAgent);
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 10]
```diff
@@ -152,6 +152,15 @@ public async void SendMutationFact()
...
         });
 
+        var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
+        var userAgent = StarWarsClient.Options.DefaultUserAgentRequestHeader;
+        if (userAgent != null)
+            defaultHeaders.UserAgent.Add(userAgent);
+        var response = await StarWarsClient.SendQueryAsync(graphQLRequest, () => new { Human = new { Name = string.Empty } });
+        callbackTester.Should().HaveBeenInvokedWithPayload().Which.Headers.Should().BeEquivalentTo(defaultHeaders);
+        Assert.Null(response.Errors);
+        Assert.Equal("Luke", response.Data.Human.Name);
```
Comment (sungam3r) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:162]:
```suggestion
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 11]
```diff
@@ -152,6 +152,11 @@ public async void SendMutationFact()
...
         });
 
+        var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
+        var userAgent = StarWarsClient.Options.DefaultUserAgentRequestHeader;
+        if (userAgent != null)
+            defaultHeaders.UserAgent.Add(userAgent);
+
         Assert.Null(mutationResponse.Errors);
         Assert.Equal("Han Solo", mutationResponse.Data.createHuman.Name);
         Assert.Equal("Corellia", mutationResponse.Data.createHuman.HomePlanet);
```
Comment (sungam3r) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:162]:
```suggestion
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 12]
```diff
@@ -152,9 +152,10 @@ public async void SendMutationFact()
             }
         });
 
-        Assert.Null(mutationResponse.Errors);
-        Assert.Equal("Han Solo", mutationResponse.Data.createHuman.Name);
-        Assert.Equal("Corellia", mutationResponse.Data.createHuman.HomePlanet);
+        var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
```
Comment (sungam3r) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:155]:
```suggestion
        Assert.Null(mutationResponse.Errors);
        Assert.Equal("Han Solo", mutationResponse.Data.createHuman.Name);
        Assert.Equal("Corellia", mutationResponse.Data.createHuman.HomePlanet);
        
        var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs [hunk 13]
```diff
@@ -155,6 +155,11 @@ public async void SendMutationFact()
         Assert.Null(mutationResponse.Errors);
         Assert.Equal("Han Solo", mutationResponse.Data.createHuman.Name);
         Assert.Equal("Corellia", mutationResponse.Data.createHuman.HomePlanet);
+        
+        var defaultHeaders = StarWarsClient.HttpClient.DefaultRequestHeaders;
+        var userAgent = StarWarsClient.Options.DefaultUserAgentRequestHeader;
+        if (userAgent != null)
+            defaultHeaders.UserAgent.Add(userAgent);
```
Comment (sungam3r) [inline tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs:162]:
```suggestion
```

File: tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs (hunk 6 above)
Comment (sungam3r) [reply tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs]: @KirovAir I resolved merge conflicts and reverted changes in this test file. As @rose-a said please add new test.

Comment (KirovAir) [reply tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs]: Sorry guys I completely overlooked it. Currently on holiday until next thursday. Let me resolve it when I get back home.

Comment (sungam3r) [reply tests/GraphQL.Integration.Tests/QueryAndMutationTests/Base.cs]: @KirovAir Let's finish it and I'm glad to merge.

**Pull request closed without merging**